### PR TITLE
Use safe dialer in hooks

### PR DESCRIPTION
**Pull request by kitop**
**- Summary**
Make sure that hooks can't access reserved IP ranges

**- Test plan**
Updated tests.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: api/signup_test.go [hunk 1]
```diff
@@ -150,6 +156,8 @@ func (ts *SignupTestSuite) TestWebhookTriggered() {
 	ts.API.handler.ServeHTTP(w, req)
 	assert.Equal(http.StatusOK, w.Code)
 	assert.Equal(1, callCount)
+
+	privateIPBlocks = append([]*net.IPNet{localhost}, privateIPBlocks...)
```
Comment (mraerino) [inline api/signup_test.go]: is that needed since there is an unshift above?

File: mailer/mailme.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+			}
+
+		},
+	})
+
+	req = req.WithContext(ctx)
+	return no.inner.RoundTrip(req)
+}
+
+func SafeRountripper(trans http.RoundTripper, log logrus.FieldLogger) http.RoundTripper {
```
Comment (mraerino) [inline mailer/mailme.go]: There's a typo here

File: mailer/template.go [hunk 3]
```diff
@@ -4,14 +4,13 @@ import (
...
 	"github.com/netlify/gotrue/models"
-	"github.com/netlify/mailme"
 )
 
 // TemplateMailer will send mail and use templates from the site for easy mail styling
 type TemplateMailer struct {
 	SiteURL string
 	Config  *conf.Configuration
-	Mailer  *mailme.Mailer
+	Mailer  *MailmeMailer
```
Comment (mraerino) [inline mailer/template.go]: we might want to patch mailme so that it can take a custom http client for fetching templates

File: api/helpers.go [hunk 4]
```diff
@@ -107,3 +110,99 @@ func (a *API) getReferrer(r *http.Request) string {
...
+		if block.Contains(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+func removeLocalhostFromPrivateIPBlock() *net.IPNet {
+	var localhost *net.IPNet
+	localhost, privateIPBlocks = privateIPBlocks[0], privateIPBlocks[1:]
```
Comment (mraerino) [inline api/helpers.go]: we should not rely on localhost being the first entry...

File: api/signup_test.go (hunk 1 above)
Comment (kitop) [reply api/signup_test.go]: Nope! Good catch, forgot to remove it.

File: mailer/template.go (hunk 3 above)
Comment (kitop) [reply mailer/template.go]: Do you wanna do this now or in a different PR?

File: mailer/mailme.go (hunk 2 above)
Comment (kitop) [reply mailer/mailme.go]: Good catch! Updating (and all references)

File: api/helpers.go (hunk 4 above)
Comment (kitop) [reply api/helpers.go]: Yeah, though about it but was lazy 🙈 
Will update!

**Comment by rybit** [comment]
I think that there is a version of this Dailer in commons. Just an FYI.

**Comment by mraerino** [comment]
we can't use commons in open source projects...

File: mailer/template.go (hunk 3 above)
Comment (mraerino) [reply mailer/template.go]: not for now

**Pull request merged**